### PR TITLE
ci(core): Seperate setup, execution, and verification of performance tests

### PR DESCRIPTION
**Pull request by jpinkney-aws**
## Problem
We are currently running performance tests without waiting for the cpu to recover. This could be part of the reason why our tests could be flaky.

These were suggestions made by George on Friday

## Solution
Wait until the the cpu usage is at a reasonable (determined using previous test runs) starting cpu usage before continuing on with the tests

---

License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/shared/performance/performance.ts [hunk 1]
```diff
@@ -85,55 +85,89 @@ export class PerformanceTracker {
...
-export function performanceTest(options: TestOptions, name: string, fn: () => void | Promise<void>) {
+export function performanceTest<TSetup, TExecute>(
+    options: TestOptions,
+    name: string,
+    fn: () => PerformanceTestFunction<TSetup, TExecute>
+) {
     const testOption = options[process.platform as 'linux' | 'darwin' | 'win32']
 
-    const totalTestRuns = options.testRuns ?? 5
+    const totalTestRuns = options.testRuns ?? 10
```
[the remark this answers is not in the document]
Comment (jpinkney-aws) [reply packages/core/src/shared/performance/performance.ts:109]: George just said increase it to 10 and see what happens

File: packages/core/src/shared/performance/performance.ts [hunk 2]
```diff
@@ -85,55 +85,89 @@ export class PerformanceTracker {
...
         after(async () => {
-            const totalUserCPUUsage =
-                testRunMetrics.reduce((acc, metric) => acc + metric.userCpuUsage, 0) / testRunMetrics.length
-            const totalSystemCPUUsage =
-                testRunMetrics.reduce((acc, metric) => acc + metric.systemCpuUsage, 0) / testRunMetrics.length
-            const totalMemoryUsage =
-                testRunMetrics.reduce((acc, metric) => acc + metric.heapTotal, 0) / testRunMetrics.length
-            const totalDuration =
-                testRunMetrics.reduce((acc, metric) => acc + metric.duration, 0) / testRunMetrics.length
+            // use median since its more resistant to outliers
```
Comment (hayemaxi) [inline packages/core/src/shared/performance/performance.ts:146]: Might also consider discarding the best and worst results and then averaging the rest, more than 1 instance of extreme values may be notable.

**Review by hayemaxi — APPROVED** [review]

File: packages/core/src/shared/performance/performance.ts [hunk 3]
```diff
@@ -85,55 +85,89 @@ export class PerformanceTracker {
     }
 }
 
+interface PerformanceTestFunction<TSetup, TExecute> {
```
Comment (nkomonen-amazon) [inline packages/core/src/shared/performance/performance.ts:88]: Not a blocker for this PR but should we add a doc to explain how to use this?

**Review by nkomonen-amazon — APPROVED** [review]

**Pull request merged**